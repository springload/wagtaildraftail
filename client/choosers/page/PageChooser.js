import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Modal from './Modal';

import * as actions from './actions';
import PageChooserHeader from './PageChooserHeader';
import PageChooserSpinner from './PageChooserSpinner';
import PageChooserBrowseView from './views/PageChooserBrowseView';
import PageChooserSearchView from './views/PageChooserSearchView';
import PageChooserErrorView from './views/PageChooserErrorView';

// TODO PageChooserExternalLinkView
// TODO PageChooserEmailView

const getTotalPages = (totalItems, itemsPerPage) => Math.ceil(totalItems / itemsPerPage);

const propTypes = {
  initialParentPageId: PropTypes.any,
  browse: PropTypes.func.isRequired,
};

const defaultProps = {
  initialParentPageId: null,
};

class PageChooser extends React.Component {
  componentDidMount() {
    const { browse, initialParentPageId } = this.props;

    browse(initialParentPageId || 'root', 1);
  }

  render() {
    const {
      browse,
      error,
      isFetching,
      items,
      onPageChosen,
      pageTypes,
      parent,
      restrictPageTypes,
      search,
      totalItems,
      viewName,
      viewOptions,
    } = this.props;
    // Event handlers
    const onSearch = (queryString) => {
      if (queryString) {
        search(queryString, restrictPageTypes, 1);
      } else {
        // Search box is empty, browse instead
        browse('root', 1);
      }
    };

    const onNavigate = (page) => {
      browse(page.id, 1);
    };

    const onChangePage = (newPageNumber) => {
      // TODO add missing default?
      switch (viewName) {
      case 'browse':
        browse(viewOptions.parentPageID, newPageNumber);
        break;
      case 'search':
        search(viewOptions.queryString, restrictPageTypes, newPageNumber);
        break;
      }
    };

    // Views
    let view = null;
    // TODO add missing default?
    switch (viewName) {
    case 'browse':
      view = (
        <PageChooserBrowseView
          parentPage={parent}
          items={items}
          pageTypes={pageTypes}
          restrictPageTypes={restrictPageTypes}
          pageNumber={viewOptions.pageNumber}
          totalPages={getTotalPages(totalItems, 20)}
          onPageChosen={onPageChosen}
          onNavigate={onNavigate}
          onChangePage={onChangePage}
        />
      );
      break;
    case 'search':
      view = (
        <PageChooserSearchView
          items={items}
          totalItems={totalItems}
          pageTypes={pageTypes}
          restrictPageTypes={restrictPageTypes}
          pageNumber={viewOptions.pageNumber}
          totalPages={getTotalPages(totalItems, 20)}
          onPageChosen={onPageChosen}
          onNavigate={onNavigate}
          onChangePage={onChangePage}
        />
      );
      break;
    }

    // Check for error
    if (error) {
      view = <PageChooserErrorView errorMessage={error} />;
    }

    return (
      <Modal>
        <PageChooserHeader onSearch={onSearch} searchEnabled={!error} />
        <PageChooserSpinner isActive={isFetching}>
          {view}
        </PageChooserSpinner>
      </Modal>
    );
  }
}

PageChooser.propTypes = propTypes;
PageChooser.defaultProps = defaultProps;

const mapStateToProps = state => ({
  viewName: state.viewName,
  viewOptions: state.viewOptions,
  parent: state.parent,
  items: state.items,
  totalItems: state.totalItems,
  pageTypes: state.pageTypes,
  isFetching: state.isFetching,
  error: state.error,
});

const mapDispatchToProps = dispatch => ({
  browse: (parentPageID, pageNumber) => dispatch(actions.browse(parentPageID, pageNumber)),
  search: (queryString, restrictPageTypes, pageNumber) => dispatch(actions.search(queryString, restrictPageTypes, pageNumber)),
});

export default connect(mapStateToProps, mapDispatchToProps)(PageChooser);
