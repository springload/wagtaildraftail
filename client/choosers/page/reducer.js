const defaultState = {
  isFetching: false,
  error: null,
  parent: null,
  items: [],
  totalItems: 0,
  pageTypes: {},
  viewName: 'browse',
  viewOptions: {
    parentPageID: 'root',
    pageNumber: 1,
  },
};

export default function pageChooser(state = defaultState, { type, payload }) {
  switch (type) {
  case 'SET_VIEW':
    return Object.assign({}, state, {
      viewName: payload.viewName,
      viewOptions: payload.viewOptions,
    });

  case 'FETCH_START':
    return Object.assign({}, state, {
      isFetching: true,
      error: null,
    });

  case 'FETCH_SUCCESS':
    return Object.assign({}, state, {
      isFetching: false,
      // parent: payload.parentJson,
      parent: null,
      items: payload.itemsJson.items,
      totalItems: payload.itemsJson.meta.total_count,
      // eslint-disable-next-line no-underscore-dangle
      pageTypes: Object.assign({}, state.pageTypes, payload.itemsJson.__types),
    });

  case 'FETCH_FAILURE':
    return Object.assign({}, state, {
      isFetching: false,
      error: payload.message,
      items: [],
      totalItems: 0,
    });

  default:
    return state;
  }
}
