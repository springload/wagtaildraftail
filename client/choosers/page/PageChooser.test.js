import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';

// import * as actions from './actions';
import reducer from './reducer';

import PageChooser from './PageChooser';

const store = createStore(reducer, {}, applyMiddleware(thunkMiddleware));

describe('PageChooser', () => {
  it('renders', () => {
    expect(shallow((
      <PageChooser store={store} onModalClose={() => {}} />
    ))).toMatchSnapshot();
    expect(shallow((
      <Provider store={store}>
        <PageChooser onModalClose={() => {}} />
      </Provider>
    ))).toMatchSnapshot();
  });
});
