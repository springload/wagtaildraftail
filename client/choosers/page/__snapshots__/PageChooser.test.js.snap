// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageChooser renders 1`] = `
<PageChooser
  browse={[Function]}
  initialParentPageId={null}
  onModalClose={[Function]}
  search={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": null,
    }
  }
/>
`;

exports[`PageChooser renders 2`] = `
<Connect(PageChooser)
  onModalClose={[Function]}
/>
`;
